File: overlays/pkgs/pomo/pomo.rs
use std::io::Write;

static IN_TMUX: std::sync::OnceLock<bool> = std::sync::OnceLock::new();

macro_rules! printf {
    ($($args:tt),*) => {{
        print!("\x1b[2K");
        print!("\x1b[100D");
        print!($($args),*);
        std::io::stdout().flush().expect("flush");
    }};
}

macro_rules! notifyf {
    ($($args:tt),*) => {{
        let in_tmux = IN_TMUX.get().unwrap_or(&false);

        if *in_tmux {
            print!("\x1bPtmux;\x1b");
        }

        print!("\x1b]777;notify;pomodoro;{}\x1b\x5c", format!($($args),*));

        if *in_tmux {
            print!("\x1b\\");
        }

        print!("\x07"); // bell
        std::io::stdout().flush().expect("flush");
    }};
}

fn pomo(status: &str, d: std::time::Duration) {
    print!("\x1b[2J");
    print!("\x1b[0;0H");
    notifyf!("{}", status);
    println!("{}", status);

    let (tx, rx) = std::sync::mpsc::channel::<()>();

    let handle = std::thread::spawn(move || {
        let start = std::time::Instant::now();
        loop {
            match rx.recv_timeout(std::time::Duration::from_secs(1)) {
                Err(std::sync::mpsc::RecvTimeoutError::Timeout) => {
                    let now = std::time::Instant::now();
                    let diff = now - start;
                    if diff == std::time::Duration::from_secs(30) {
                        notifyf!("30 seconds left!");
                    }
                    let time_left = d - diff;
                    let time_left_display = if time_left > std::time::Duration::from_secs(60) {
                        format!("{}min", time_left.as_secs().div_ceil(60))
                    } else {
                        format!("{}sec", time_left.as_secs())
                    };
                    printf!("{}", time_left_display);
                }
                _ => break,
            }
        }
    });

    std::thread::sleep(d);
    tx.send(()).expect("could not send to thread");
    handle.join().expect("could not join thread handle");
}

fn main() {
    _ = IN_TMUX.get_or_init(|| std::env::var("TMUX").is_ok());

    loop {
        for _ in 0..4 {
            pomo("work!", std::time::Duration::from_secs(25 * 60));
            pomo("break!", std::time::Duration::from_secs(5 * 60));
        }
        pomo("long break!", std::time::Duration::from_secs(30 * 60));

        println!("Press <ENTER> to continue to the next pomodoro session, CTRL-C to quit.");
        let mut lines = std::io::stdin().lines();
        _ = lines.next();
    }
}

File: overlays/by-name/pomo/pomo.zig
const std = @import("std");

var inside_tmux = false;

var out_buf = [_]u8{0} ** 1024;
var stdout_file = std.fs.File.stdout().writer(&out_buf);

fn notify(message: []const u8) !void {
    var stdout = &stdout_file.interface;
    try stdout.print("{s}\n", .{message});

    if (inside_tmux) {
        try stdout.writeAll("\x1bPtmux;\x1b");
    }

    try stdout.print("\x1b]777;notify;pomodoro;{s}\x1b\x5c", .{message});

    if (inside_tmux) {
        try stdout.writeAll("\x1b\\");
    }

    try stdout.writeAll("\x07"); // bell
    try stdout.flush();
}

fn pomo(
    message: []const u8,
    duration: comptime_int,
) !void {
    var stdout = &stdout_file.interface;
    try stdout.writeAll("\x1b[2J\x1b[0;0H");
    try stdout.flush();
    try notify(message);

    const warning_time_ns = @min(duration / 10, 30 * std.time.ns_per_s);

    var msg_buf = [_]u8{0} ** 30;
    const msg = try std.fmt.bufPrint(&msg_buf, "{}s left!", .{@divFloor(warning_time_ns, std.time.ns_per_s)});

    std.Thread.sleep(duration - warning_time_ns);
    try notify(msg);
    std.Thread.sleep(warning_time_ns);
}

pub fn main() !void {
    var arena = std.heap.ArenaAllocator.init(std.heap.page_allocator);
    defer arena.deinit();

    const allocator = arena.allocator();

    var stdin_file = std.fs.File.stdin().reader(&.{});
    var stdin = &stdin_file.interface;

    inside_tmux = b: {
        _ = std.process.getEnvVarOwned(allocator, "TMUX") catch |err| switch (err) {
            error.EnvironmentVariableNotFound => break :b false,
            else => return err,
        };

        break :b true;
    };

    while (true) {
        for (0..4) |_| {
            try pomo("work!", 25 * std.time.ns_per_min);
            try pomo("break!", 5 * std.time.ns_per_min);
        }

        try pomo("long break!", 30 * std.time.ns_per_min);

        std.debug.print(
            "Press <ENTER> to continue to the next pomodoro session, CTRL-C to quit.",
            .{},
        );

        // We don't have the controlling terminal in raw mode, so input will
        // be buffered until a newline is entered. This means if we succeed to
        // read anything, then the user hit <ENTER>.
        _ = try stdin.takeByte();
    }
}
